pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,380 @@
+	// DefaultKubeconfigContext is the default context to use inside the kubeconfig file. None means use the current
+	// context without switching.
+	DefaultKubeconfigContext = ""
+	// DefaultKubeconfigPath is the default path to the kubeconfig file. None means use in-cluster auth.
+	DefaultKubeconfigPath = ""
+	// DefaultResyncPeriod is the default resync period for the pod cache.
+	DefaultResyncPeriod = 5 * time.Minute
+	// DefaultWatchCluster is the default watch mode for pods. By default we watch the entire cluster.
+	DefaultWatchCluster = true
+)
+
+// Provider represents a k8s provider.
+type Provider struct {
+	logger logrus.FieldLogger
+
+	PodsInf             cache.SharedIndexInformer
+	annotationWhitelist []regexp.Regexp
+	labelWhitelist      []regexp.Regexp
+}
+
+func (p *Provider) EstimatedTags() int {
+	// There is no real way to estimate this for k8s provider as any pod can have arbitrary labels/annotations
+	return 8
+}
+
+// Instance returns pod details from k8s API server watches.
+// ip -> nil pointer if pod was not found.
+// map is returned even in case of errors because it may contain partial data.
+// An "instance", as far as k8s cloud provider is concerned, is an individual pod running in the cluster
+func (p *Provider) Instance(ctx context.Context, IP ...gostatsd.IP) (map[gostatsd.IP]*gostatsd.Instance, error) {
+	instanceIPs := map[gostatsd.IP]*gostatsd.Instance{}
+	var returnErr error
+
+	// Lookup via the pod cache
+	for _, lookupIP := range IP {
+		if lookupIP == gostatsd.UnknownIP {
+			continue
+		}
+
+		p.logger.WithField("ip", lookupIP).Debug("Looking up pod ip")
+		objs, err := p.PodsInf.GetIndexer().ByIndex(PodsByIPIndexName, string(lookupIP))
+		if err != nil {
+			instanceIPs[lookupIP] = nil
+			returnErr = err
+			continue
+		}
+		if len(objs) < 1 {
+			p.logger.Debugf("Could not find IP in cache, continuing")
+			instanceIPs[lookupIP] = nil
+			continue
+		}
+		if len(objs) > 1 {
+			p.logger.Warn("More than one pod in cache for IP '%s'. Using first stored", lookupIP)
+		}
+		pod := objs[0].(*core_v1.Pod)
+
+		// Turn the pod metadata into tags
+		tags := gostatsd.Tags{}
+		// TODO: deduplicate labels and annotations in their tag format, rather than overwriting
+		for k, v := range pod.ObjectMeta.Labels {
+			if tagKeyMatchesWhitelist(k, p.labelWhitelist) {
+				tags = append(tags, fmt.Sprintf("%s:%s", k, v))
+			}
+		}
+		for k, v := range pod.ObjectMeta.Annotations {
+			if !tagKeyMatchesWhitelist(k, p.annotationWhitelist) {
+				continue
+			}
+			// Annotations are often of the format "purpose.company.com/annotationName" so we want to expectedKey off the
+			// start when processing tag names
+			tagName := k
+			splitTagName := strings.SplitN(k, "/", 2)
+			if len(splitTagName) > 1 {
+				tagName = splitTagName[1]
+			}
+			tags = append(tags, fmt.Sprintf("%s:%s", tagName, v))
+		}
+		instanceID := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+		instanceIPs[lookupIP] = &gostatsd.Instance{
```

Comment:
Not in this PR but I wonder if we should change this whole "API" to use values and use the `ID` field as the flag? Making some garbage for the collector to collect here...

Reply:
Yeah I think that would fit in with the cache changes, or close to.